app/services/analytics_service.rb

```diff
@@ -8,110 +8,179 @@ def initialize(user_or_org, start_date: "", end_date: "", article_id: nil)
     load_data
   end
 
+  # Totals computes total counts for comments, reactions, follows and page views
   def totals
-    total_views = article_data.sum(:page_views_count)
-    logged_in_page_view_data = page_view_data.where.not(user_id: nil)
-    average_read_time_in_seconds = average_read_time(logged_in_page_view_data)
-
     {
-      comments: {
-        total: comment_data.size
-      },
-      reactions: {
-        total: reaction_data.size,
-        like: reaction_data.count { |rxn| rxn.category == "like" },
-        readinglist: reaction_data.count { |rxn| rxn.category == "readinglist" },
-        unicorn: reaction_data.count { |rxn| rxn.category == "unicorn" }
-      },
-      follows: {
-        total: follow_data.size
-      },
-      page_views: {
-        total: total_views,
-        average_read_time_in_seconds: average_read_time_in_seconds,
-        total_read_time_in_seconds: average_read_time_in_seconds * total_views
-      }
+      comments: { total: comment_data.size },
+      follows: { total: follow_data.size },
+      reactions: calculate_reactions_totals,
+      page_views: calculate_page_views_totals
     }
   end
 
-  def stats_grouped_by_day
-    result = {}
-
-    (start_date.to_date..end_date.to_date).each do |date|
-      result[date.to_s(:iso)] = cached_data_for_date(date)
+  # Grouped by day computes counts for comments, reactions, follows and page views per each day
+  def grouped_by_day
+    return {} unless start_date && end_date
+
+    # cache all stats in the date range for the requested user or organization
+    cache_key = "analytics-for-dates-#{start_date}-#{end_date}-#{user_or_org.class.name}-#{user_or_org.id}"
+
+    Rails.cache.fetch(cache_key, expires_in: 7.days) do
```

Comment:
Since now there's only one cache per range, I kept it to the upper range, 7 days, but we can probably lower it to 3 if we see that API clients aren't going to keep answering the same range over and over

Reply:
What would happen if someone requested the past seven days? That would cache the most recent day for a long time right?

Reply:
I changed the cache key to include both start and end date, so if the past seven days are asked twice the same day the second time they come back from the cache. The next day the cache key changes. I believe it's a good compromise but I'm not 100% sure it's the right compromise.

This would help dashboards with fixed dates without killing performance by caching each day separately.

I think we can revisit the caching policy in the future because it really depends on how all of this behaves in production once we have enough data to understand which caching strategy would work best.

Caching is hard, no doubt about that ehhe